script/model/model_server.py

```diff
@@ -0,0 +1,355 @@
+The message format that the ModelServer expects should be kept consistent with Messenger class in
+the noisepage source code.
+The command format should be kept with ModelServerManager in the noisepage source
+code.
+
+TODO(Ricky):
+- Encapsulate the ModelServer in a class
+
+"""
+
+from __future__ import annotations
+import sys
+import atexit
+from enum import Enum, auto, IntEnum
+from typing import Dict, Optional, Tuple, List, Any
+import json
+import logging
+import pprint
+import pickle
+import numpy as np
+import zmq
+
+from data_class import opunit_data
+from mini_trainer import MiniTrainer
+from sklearn import model_selection
+from util import logging_util
+from type import OpUnit
+
+
+logging_util.init_logging('info')
+
+
+
+
+class Callback(IntEnum):
+    """
+    ModelServerManager <==> ModelServer callback Id.
+    Needs to be kept consistent with ModelServerManager.h's Callback Enum
+    """
+    NOOP = 0
+    CONNECTED = 1
+
+class Command(Enum):
+    """
+    Command enum for actions to take from the manager.
+    This has to be kept consistent with the C++ ModelServerManager.
+    """
+    TRAIN = auto()      # Train a specific model
+    QUIT = auto()       # Quit the server
+    PRINT = auto()      # Print the message
+    INFER = auto()      # Do inference on a trained model
+
+    def __str__(self) -> str:
+        return self.name
+
+    @staticmethod
+    def from_str(cmd_str: str) -> Command:
+        if cmd_str == "PRINT":
+            return Command.PRINT
+        elif cmd_str == "QUIT":
+            return Command.QUIT
+        elif cmd_str == "TRAIN":
+            return Command.TRAIN
+        elif cmd_str == "INFER":
+            return Command.INFER
+        else:
+            raise ValueError("Invalid command")
+
+
+
+class Message:
+    """
+    Message struct for communication with the ModelServer.
+    The message format has to be ketp consistent with the C++ Messenger.
+    A valid message is :
+        "send_id-recv_id-payload"
+
+    Refer to Messenger's documention for the message format
+    """
+    def __init__(self, cmd: Optional[Command] = None, data: Optional[Dict] = None) -> None:
+        self.cmd = cmd
+        self.data = data
+
+    @staticmethod
+    def from_json(json_str: str) -> Message:
+        d = json.loads(json_str)
+        msg = Message()
+        try:
+            msg.cmd = Command.from_str(d["cmd"])
+            msg.data = d["data"]
+        except (KeyError, ValueError) as e:
+            logging.error(f"Invalid Message : {json_str}")
+            return None
+
+        return msg
+
+    def to_json(self) -> str:
+        return json.dumps(self.__dict__)
+
+    def __str__(self) -> str:
+        return pprint.pformat(self.__dict__)
+
+
+class ModelServer:
+    """
+    ModelServer(MS) class that runs in a loop to handle commands from the ModelServerManager from C++
+    """
+
+    # Training parameters
+    RESULT_PATH = '/tmp'
```

Comment:
I think this should be managed by the c++ side. Both the `train` and `infer` command should include this field as a parameter. `train` only saves the models on disk at the specified path. And `infer` checks whether the models from the specified path have been loaded (cached). If not, load the models from the path.

Reply:
Oh, I think I might have made false assumptions about the use case. So the models trained would be persistent across model server lifetimes? Like the pilot would train the model A, and then expects to re-use it after the model server reboots?
I was assuming the models will be trained every ModelServer's lifetime, like train->infer->infer->infer, and nothing is persisted. 
But yeah, I could have a persisted model abstraction with filename as the handle.

Reply:
Yes. The trained models would expect to be persistent across model server lifetimes.